File: src/main/java/edu/cornell/cs/apl/viaduct/security/Principal.java
package edu.cornell.cs.apl.viaduct.security;

import com.google.auto.value.AutoValue;

/** An actor with an associated security label. */
@AutoValue
public abstract class Principal implements Comparable<Principal> {
  public static Principal create(String name) {
    return new AutoValue_Principal(name);
  }

  abstract String getName();

  @Override
  public final int compareTo(Principal that) {
    return this.getName().compareTo(that.getName());
  }

  @Override
  public final String toString() {
    return getName();
  }
}

File: compiler/src/main/kotlin/edu/cornell/cs/apl/viaduct/security/Principal.kt
package edu.cornell.cs.apl.viaduct.security

import edu.cornell.cs.apl.prettyprinting.Document
import edu.cornell.cs.apl.prettyprinting.PrettyPrintable
import edu.cornell.cs.apl.prettyprinting.Style
import edu.cornell.cs.apl.prettyprinting.styled

/** An actor with an associated security label. */
data class Principal(val name: String) : Comparable<Principal>, PrettyPrintable {
    override fun compareTo(other: Principal): Int =
        name.compareTo(other.name)

    override val asDocument: Document
        get() = Document(name).styled(PrincipalStyle)

    // TODO: remove and use [asDocument]
    override fun toString(): String =
        name
}

/** The display style of [Principal]s. */
object PrincipalStyle : Style
